manul/src/session/message.rs

```diff
@@ -324,3 +348,60 @@ impl<Verifier> VerifiedMessage<Verifier> {
         (echo_broadcast, normal_broadcast, direct_message)
     }
 }
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) struct SignedMessageHash {
+    signature: SerializedSignature,
+    metadata: MessageMetadata,
+    #[serde(with = "SliceLike::<Hex>")]
+    payload_hash: Box<[u8]>,
+}
+
+impl SignedMessageHash {
+    pub(crate) fn metadata(&self) -> &MessageMetadata {
+        &self.metadata
+    }
+
+    pub(crate) fn verify<SP>(self, verifier: &SP::Verifier) -> Result<VerifiedMessageHash, MessageVerificationError>
+    where
+        SP: SessionParameters,
+    {
+        let digest =
+            message_digest::<SP>(&self.metadata, &self.payload_hash).map_err(MessageVerificationError::Local)?;
+        let signature = self
+            .signature
+            .deserialize::<SP>()
+            .map_err(|_| MessageVerificationError::InvalidSignature)?;
+        if verifier.verify_digest(digest, &signature).is_ok() {
+            Ok(VerifiedMessageHash {
+                signature: self.signature,
+                metadata: self.metadata,
+                payload_hash: self.payload_hash,
+            })
+        } else {
+            Err(MessageVerificationError::SignatureMismatch)
+        }
```

Comment:
```suggestion
        verifier
            .verify_digest(digest, &signature)
            .map(|r| VerifiedMessageHash {
                signature: self.signature,
                metadata: self.metadata,
                payload_hash: self.payload_hash,
            })
            .map_err(|_e| MessageVerificationError::SignatureMismatch)
```

If we had a `From<signature::Error> for MessageVerificationError` impl the last line could go away too.